src/disposer/AsyncDisposer.ts

```diff
@@ -0,0 +1,64 @@
+import { IAsyncDisposable } from './types';
+
+const listeners = new WeakMap<DisposeFn, Set<DisposeFn>>();
+
+type DisposeFn = () => void | Promise<void>;
+
+declare class AsyncDisposer implements IAsyncDisposable {
+  constructor(dispose: DisposeFn);
+  public dispose(): void | Promise<void>;
+  public readonly disposed: boolean;
+  public add(dispose: () => void): void;
+}
+
+function AsyncDisposer(dispose: DisposeFn): AsyncDisposer {
+  let _dispose: DisposeFn | undefined = dispose;
+  const self: AsyncDisposer = Object.defineProperties(
+    {},
+    {
+      dispose: {
+        enumerable: true,
+        configurable: true,
+        async value() {
+          if (!_dispose) {
+            console.warn('Already disposed');
+          } else {
+            await _dispose();
+            const callbacks = listeners.get(_dispose);
+            if (callbacks) {
+              await Promise.all(
```

Comment:
Not supported in enough browsers methinks. Also, what could possibly throw XD

Reply:
Actually seriously, expand on that idea. What do you think should happens if things throw? Do we re-throw the result from `allSettled`?

Reply:
>  Do we re-throw the result from allSettled?

Yeah, we re-throw if any of 'settled' promise has rejected status.

Reply:
I'm hesitant to add anything to `lifecycle` that requires polyfilling, since we still don't have a polyfilling strategy.

Even if we restrict `@stoplight/lifecycle` to only Studio, it doesn't meet the browser requirements for `Promise.allSettled`:

![image](https://user-images.githubusercontent.com/587740/91203736-4ba6d700-e6d1-11ea-978b-6ea0a0c2762d.png)

![image](https://user-images.githubusercontent.com/587740/91203842-5eb9a700-e6d1-11ea-88a7-81998c48f20a.png)

Plus, [according to MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled#Implementation_Progress), neither Chrome nor Firefox's implementations actually pass all the TEST262 test suite yet. So I hesitate to use it.
![image](https://user-images.githubusercontent.com/587740/91204389-03d47f80-e6d2-11ea-99c0-e386fcb03116.png)


Secondly, it's not necessary to use `allSettled` to prevent memory leaks. All of the `.dispose()` methods will have been called, so the only advantage of `allSettled` is if TWO or more `dispose` methods throw, we'll see ALL of them reported. But that leads to my third reason..

We have no upstream code (say in `@stoplight/reporter`) designed to handle [AggregateError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError) from `Promise.any` much less the funny array of `{ status, reason|value }` objects returned by `Promise.all`. You could argue it's a chicken and egg problem, but do we really care about all of the errors?

If you think we care, I'll update this code with a re-implementation of Promise.allSettled so it works in all browsers and throws an `AggregateError`. Actually.... yeah I'll do that. I think we care.

But we still shouldn't use `Promise.allSettled`. :trollface: